test/Olcs/src/FormService/Form/Lva/ApplicationFinancialEvidenceTest.php

```diff
@@ -43,106 +43,18 @@ class ApplicationFinancialEvidenceTest extends MockeryTestCase
-                        ]
-                    );
-                    return $element;
-                }
-            );
-
-            $serviceManager->get('FormElementManager')->setFactory(
-                'Common\Form\Element\DynamicMultiCheckbox',
-                function ($serviceLocator, $name, $requestedName) {
-                    $element = new DynamicMultiCheckbox();
-                    $element->setValueOptions(
-                        [
-                            '1' => 'one',
-                            '2' => 'two',
-                            '3' => 'three'
-                        ]
-                    );
-                    return $element;
-                }
-            );
-
-        return $serviceManager;
-    }
-
-    /**
-     * Added this method for backwards compatibility
-     *
-     * @return \Laminas\ServiceManager\ServiceManager
-     */
-    public static function getRealServiceManager()
-    {
-        $serviceManager = new ServiceManager(new ServiceManagerConfig());
-        $config = include 'config/application.config.php';
-        $serviceManager->setService('ApplicationConfig', $config);
-        $serviceManager->get('ModuleManager')->loadModules();
-        $serviceManager->setAllowOverride(true);
-
-        $mockTranslationLoader = m::mock(TranslationLoader::class);
-        $mockTranslationLoader->shouldReceive('load')->andReturn(['default' => ['en_GB' => []]]);
-        $mockTranslationLoader->shouldReceive('loadReplacements')->andReturn([]);
-        $serviceManager->setService(TranslationLoader::class, $mockTranslationLoader);
-
-        $pluginManager = new LoaderPluginManager($serviceManager);
-        $pluginManager->setService(TranslationLoader::class, $mockTranslationLoader);
-        $serviceManager->setService('TranslatorPluginManager', $pluginManager);
-
-        // Mess up the backend, so any real rest calls will fail
-        $config = $serviceManager->get('Config');
-        $config['service_api_mapping']['endpoints']['backend'] = 'http://some-fake-backend/';
-        $serviceManager->setService('Config', $config);
-
-        return $serviceManager;
-    }
-
     public function setUp(): void
     {
         $this->fh = m::mock(FormHelperService::class)->makePartial();
         $this->fsm = m::mock(\Common\FormService\FormServiceManager::class)->makePartial();
         $this->urlHelper = m::mock(UrlHelperService::class);
         $this->translator = m::mock(TranslationHelperService::class);
 
-        $sm = $this->getServiceManager();
-        $sm->setService('Helper\Url', $this->urlHelper);
-        $sm->setService('Helper\Translation', $this->translator);
 
-        $this->fsm->shouldReceive('getServiceLocator')->andReturn($sm);
+        $this->fsm->setService('Helper\Url', $this->urlHelper);
+        $this->fsm->setService('Helper\Translation', $this->translator);
+
+        $this->fsm->shouldReceive('getServiceLocator')->andReturn($this->fsm);
```

Comment:
How come we're mocking these now on `FormServiceManager` instead of mocking a `ContainerInterface`?

Do `Helper\Url` or `Helper\Translation` get used?

Reply:
**Explicit Dependencies:**
Mocking specific services on `FormServiceManager` makes test more explicit about the dependencies of class `ApplicationFinancialEvidenceTest` . 
Here we can exactly mentioned which services class depends on, and make test more readable and maintainable

**ContainerInterface Abstraction**
Might introduce unnecessary complexity if  class requires a specific set of services. it allows for more flexiblity but may not accurately reflect the actual dependencies of class.

**Proposed solution:**
```    public function setUp(): void
    {
        $this->fh = m::mock(FormHelperService::class)->makePartial();
        $this->urlHelper = m::mock(UrlHelperService::class);
        $this->translator = m::mock(TranslationHelperService::class);

        $this->fsm = m::mock(\Common\FormService\FormServiceManager::class)->makePartial();
        $this->fsm->shouldReceive('getServiceLocator')->andReturn($this->fsm);

        $this->fsm->setService('Table', $this->urlHelper);
        $this->fsm->setService('Helper\Translation', $this->translator);

        $this->sut = new ApplicationFinancialEvidence($this->fh, m::mock(AuthorizationService::class), $this->translator, $this->urlHelper);
    }```

Reply:
> Explicit Dependencies:
Mocking specific services on FormServiceManager makes test more explicit about the dependencies of class ApplicationFinancialEvidenceTest .
Here we can exactly mentioned which services class depends on, and make test more readable and maintainable

Tests are testing the source code and should mirror what's happening there. Is the source code re-using the same service manager or setting a new one?

Also, did you check if `Helper\Url` or `Helper\Translation` get retrieved ?

> ContainerInterface Abstraction
Might introduce unnecessary complexity if class requires a specific set of services. it allows for more flexibility but may not accurately reflect the actual dependencies of class.

`ContainerInterface` is a standardised interface ([PSR-11](https://www.php-fig.org/psr/psr-11/)). I'd argue it's the opposite of complexity.

> [!NOTE]  
> May need to use old container interface depending on what has migrated to PSR-11. https://github.com/container-interop/container-interop